### ISSUE TITLE
`read_csv` significantly slower than pandas when passing in a file handle

### ISSUE THREAD

**Issue by mzjp2** [modin-project/modin#2624]
Hello! I've done a bunch of research around "modin" and "file handle" and couldn't find anything, so I'm opening this issue.

I'm using the `DM_ALUNO.csv` (about 2.9G in size with 12m rows) in an effort to benchmark `modin` and `pandas` from this URL `http://download.inep.gov.br/microdados/microdados_educacao_superior_2018.zip`.

I'm running on a 13" Macbook Pro 2019 with 16gb RAM and a 2.8 GHz Quad-Core Intel Core i7 CPU. `modin` (with the dask backend) performs faster than Pandas consistently by about 25% when doing:

```
import modin.pandas as pd

pd.read_csv("DM_ALUNO.CSV")
```

on my computer, the above takes about 17s on `modin` and 25s on `pandas`. However, if I pass in a file handle instead:

```
import modin.pandas as pd

with open("DM_ALUNO.CSV") as f:
  pd.read_csv(f)
```

then it performs _much_ worse. It takes about 60s on `modin` and continues to take 25s on `pandas`. Does anybody know why this is happening? I had a browse through the `read_csv` code on `modin` and couldn't fathom why. Does this have something to do with `dask` and it being unable to pickle file handlers?

Note, I also get the following when doing this:

```
>>> with open("data/01_raw/DM_ALUNO.CSV") as f:
...   import modin.pandas as pd; start_time=time.time();pd.read_csv(f);end_time=time.time()
...
UserWarning: Parameters provided defaulting to pandas implementation.
To request implementation, send an email to feature_requests@modin.org.
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 1.52 GB -- Worker memory limit: 2.15 GB
distributed.worker - WARNING - Memory use is high but worker has no data to store to disk.  Perhaps some other process is leaking memory?  Process memory: 1.53 GB -- Worker memory limit: 2.15 GB
```

**Comment by mzjp2** [comment]
hmm, okay - so from what I can see in `csv_dispatcher.py` - you revert to `pandas` if what's passed in is a file handle (in fact, the only time you don't revert to pandas is if it's a `pypath/pathlib` or a string that references a path that exists (on S3 or locally)

**Comment by devin-petersohn** [comment]
>Do you know why? And even then, why is falling back to pandas roughly 3x as slow as using pandas itself?

File handles have state in them that cannot be shared across workers. What we do to avoid having shared file handles is to have each worker create its own file handle from the filename. 

When falling back to pandas, the data must be distributed after its read in, this means communication large objects between workers. Even in a single machine this can be extremely expensive. It is bottlenecked by how quickly the compute engine pickles and communicates data. 

Your fix in #2625 is great!
